lxd/db/snapshots.mapper.go

```diff
@@ -86,6 +112,21 @@ var instanceSnapshotRename = cluster.RegisterStmt(`
 UPDATE instances_snapshots SET name = ? WHERE instance_id = (SELECT instances.id FROM instances JOIN projects ON projects.id = instances.project_id WHERE projects.name = ? AND instances.name = ?) AND name = ?
 `)
 
+var instanceSnapshotDeleteByInstance = cluster.RegisterStmt(`
+DELETE FROM instances_snapshots WHERE instance_id = (SELECT instances.id FROM instances WHERE instances.name = ?)
+`)
+var instanceSnapshotDeleteByProjectAndInstance = cluster.RegisterStmt(`
+DELETE FROM instances_snapshots WHERE instance_id = (SELECT instances.id FROM instances JOIN projects ON projects.id = instances.project_id WHERE projects.name = ? AND instances.name = ?)
+`)
+var instanceSnapshotDeleteByName = cluster.RegisterStmt(`
+DELETE FROM instances_snapshots WHERE name = ?
+`)
+var instanceSnapshotDeleteByProjectAndName = cluster.RegisterStmt(`
+DELETE FROM instances_snapshots WHERE name = ?
+`)
```

Comment:
> Also @masnax this PR reminds me of [#8579 (comment)](https://github.com/lxc/lxd/pull/8579#issuecomment-800958211) and I wondered as you're looking at the DB generator now, do your changes effectively render that scenario impossible to occur now? Thanks

Right, I believe since we're generating every possible db statement for the Filter structs, we won't run into that specific issue with missing statements anymore.

That being said, this line here is still a bit awkward. There's still the potential for misunderstandings arising from indirect references. `Project` in this line requires `Instance` to also be passed. Hence the actual filter is just checking `Name`. 

I'll make a small update that omits these combinations from the power set.

Reply:
Again, forgive me if I'm missing something obvious, as I am not really familiar with the DB generation code. But in what scenario would a query like `DELETE FROM instances_snapshots WHERE name = ?` be useful or safe to run? This seems to be asking for accidental deletion of snapshots that share the same name across multiple instances?

As the instances_snapshots `name` column does not include the parent instance name, nor does it include the project.

e.g.
```
+----+-------------+-------+-------------------------------------+----------+-------------+---------------------------+
| id | instance_id | name  |            creation_date            | stateful | description |        expiry_date        |
+----+-------------+-------+-------------------------------------+----------+-------------+---------------------------+
| 60 | 400         | snap0 | 2021-07-20T17:06:23.913948836+01:00 | 0        |             | 0000-12-31T23:58:45-00:01 |
+----+-------------+-------+-------------------------------------+----------+-------------+---------------------------+
```

I'm pretty sure we don't want to have those types of queries generated as they are asking for trouble :)

Reply:
Do we have a mechanism to only generate combinations for fields that have to be supplied for a filter, and then error out if they are not provided?

Reply:
@tomponline Regarding the initial question, the issue in 8579 is what motivated some of this indeed.
Basically @masnax is learning about the DB generation code as it's an area where we were lacking expertise and where we never quite completed the migration.

The goal is to look at tables which we currently handled manually and move them over to using generated code.
And as part of that, we also want to reduce the risk of human error further by removing the need for telling the generator what queries we want.

I suspect there are alternative approaches which would work just fine too. One I remember suggesting was to keep things the way we have them but turn all the filter fields into pointers and have the filtering logic outright fail (error/panic) if there are fields in the filter struct which weren't consumed by the generated query (so catching cases like in 8579).

As for memory usage, I'm not sure. It feels like we should at least make sure they're `const` which in theory would then just increase the binary size but not come with much of a runtime cost?